### PR TITLE
"edit.h/.cc" splitted to limit rebuilding dependencies

### PR DESCRIPTION
**Pull request by Hombre57**
@pandagrapher @Desmis @Beep6581 

I've splitted the `edit.h/.cc` file because it was rebuilding too much dependent files in the `spot-removal-tool` branch. I'd like to commit this branch ASAP to continue the work on the spot removal feature.

If you've added things in the edit file, I'd recommend to make a backup of the actual one, because the merge will be messy. `Geometry` subclasses will have to go in the `editwidgets` files.

I'm not proud of the `editcallbacks` file name, but I couldn't find a better and still representative one.

I'll merge this branch tomorrow if no one complain.

### CODE REVIEW

File: rtgui/editcallbacks.h [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+class EditSubscriber
+{
+
+public:
+
+private:
+    EditUniqueID ID; /// this will be used in improcfun to locate the data that has to be stored in the buffer; it must be unique in RT
+    EditType editingType;
+    BufferType bufferType;
+    EditDataProvider *provider;
```
Comment (Floessie) [inline rtgui/editcallbacks.h:42]: Who owns `provider`? If it's `EditSubscriber` and `getEditProvider()` has to return a mutable instance then better use a `std::shared_ptr<>`.

If the provider isn't owned by `EditSubscriber` but already exists when the subscriber is created, why not provide it in the c'tor and store a `const &`?

**Comment by Pandagrapher** [comment]
@Desmis
I can do the merge for newlocallab branch this WE
Pierre

File: rtgui/editcallbacks.h [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+
+protected:
+    std::vector<Geometry*> visibleGeometry;    /// displayed geometry
```
Comment (Floessie) [inline rtgui/editcallbacks.h:45]: Again, who owns `Geometry*`? Maybe a `std::unique_ptr<>` is due.

File: rtgui/editcallbacks.h [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+    std::vector<Geometry*> mouseOverGeometry;  /// mouseOver geometry, drawn in a hidden buffer
+    enum {
```
Comment (Floessie) [inline rtgui/editcallbacks.h:47]:
```c++
enum class Action {
    NONE,
    DRAGGING,
    PICKING
};

Action action;
```

This makes for better names. See `toUnderlying()` for explicit conversion.

File: rtgui/editcallbacks.h [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+        ES_ACTION_NONE,      ///
+        ES_ACTION_DRAGGING,  /// set action to this value in the buttonPressed event to start dragging and ask for drag event
+        ES_ACTION_PICKING    /// set action to this value in the buttonPressed event whenever the user is picking something through a single click. In this case, the pickX events will be called INSTEAD of buttonXReleased !
+    } action;                /// object mode only, ignored in Pipette mode
+
+public:
+    explicit EditSubscriber (EditType editType);
+    virtual ~EditSubscriber () {}
```
Comment (Floessie) [inline rtgui/editcallbacks.h:55]:
```c++
virtual ~EditSubscriber() = default;
```

File: rtgui/editcallbacks.h [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+
+    void               setEditProvider(EditDataProvider *provider);
```
Comment (Floessie) [inline rtgui/editcallbacks.h:57]: Might be unneeded if passing through c'tor is possible.

File: rtgui/editcallbacks.h [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+    EditDataProvider*  getEditProvider ();
```
Comment (Floessie) [inline rtgui/editcallbacks.h:58]: Getters should  be `const`.

Is a mutable provider (returned) necessary?

File: rtgui/editcallbacks.h [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+    void               setEditID(EditUniqueID ID, BufferType buffType);
+    bool               isCurrentSubscriber();
```
Comment (Floessie) [inline rtgui/editcallbacks.h:60]: `const`.

File: rtgui/editcallbacks.h [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+    virtual void       subscribe();
+    virtual void       unsubscribe();
+    virtual void       switchOffEditMode ();    /// Occurs when the user want to stop the editing mode
+    EditUniqueID       getEditID();
```
Comment (Floessie) [inline rtgui/editcallbacks.h:64]: `const`. Also applies to the next four lines.

File: rtgui/editcallbacks.h [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+    EditType           getEditingType();
+    BufferType         getPipetteBufferType();
+    bool               isDragging();            /// Returns true if something is being dragged and drag events has to be sent (object mode only)
+    bool               isPicking();             /// Returns true if something is being picked
+
+    /** @brief Get the cursor to be displayed when above handles
+    @param objectID object currently "hovered" */
+    virtual CursorShape getCursor (const int objectID);
```
Comment (Floessie) [inline rtgui/editcallbacks.h:72]: IMHO you should either inline the default behavior here directly or move the implementation to the `.cc`. Splitting it to two places in the header sometimes makes sense for template implementation but not so much for standard methods.

The `const int objectID` should really be `int objectID` as this is a top level `const` in a declaration. See my comment and the links [here](https://github.com/CarVac/librtprocess/issues/29#issuecomment-461724660).

File: rtgui/editcallbacks.h [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+    virtual bool pick2 (const bool picked);
+
+    /** @brief Triggered when the user is releasing mouse button 3 while in action==ES_ACTION_PICKING mode
+    @param picked True if the cursor is still above the the same object than on button pressed and with the same modifier keys.
+                  If false, the user moved the cursor away or the modifier key is different, so the element is considered as NOT selected.
+    @return true if the preview has to be redrawn, false otherwise */
+    virtual bool pick3 (const bool picked);
+
+    /** @brief Get the geometry to be shown to the user */
+    const std::vector<Geometry*>& getVisibleGeometry ();
```
Comment (Floessie) [inline rtgui/editcallbacks.h:145]: Is the caller allowed to change the geometries? Does the calling change the `EditSubscriber` instance? If the answer is "no" twice consider this:

```c++
std::vector<const Geometry*> getVisibleGeometry() const;
```

File: rtgui/editcallbacks.h [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+ * trigger the usual event so that the image will be reprocessed to compute the buffer of the current subscriber.
+ */
+class EditDataProvider
+{
+
+private:
+    EditSubscriber *currSubscriber;
+
+public:
+    int object;            /// ET_OBJECTS mode: Object detected under the cursor, 0 otherwise; ET_PIPETTE mode: 1 if above the image, 0 otherwise
```
Comment (Floessie) [inline rtgui/editcallbacks.h:163]: Public data breaks encapsulation. You could put it into an inner `struct` and add getters/setters.

File: rtgui/editcallbacks.h [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+
+    rtengine::Coord posScreen;       /// Location of the mouse button press, in preview image space
+    rtengine::Coord posImage;        /// Location of the mouse button press, in the full image space
+    rtengine::Coord deltaScreen;     /// Delta relative to posScreen
+    rtengine::Coord deltaImage;      /// Delta relative to posImage
+    rtengine::Coord deltaPrevScreen; /// Delta relative to the previous mouse location, in preview image space
+    rtengine::Coord deltaPrevImage;  /// Delta relative to the previous mouse location, in the full image space
+
+    EditDataProvider();
+    virtual ~EditDataProvider() {}
```
Comment (Floessie) [inline rtgui/editcallbacks.h:174]: `= default`.

File: rtgui/editcallbacks.h [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+
+    virtual void        subscribe(EditSubscriber *subscriber);
+    virtual void        unsubscribe();         /// Occurs when the subscriber has been switched off first
+    virtual void        switchOffEditMode ();  /// Occurs when the user want to stop the editing mode
+    virtual CursorShape getCursor(int objectID);
```
Comment (Floessie) [inline rtgui/editcallbacks.h:179]: `const`.

File: rtgui/editcallbacks.h [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+    int                 getPipetteRectSize ();
+    EditSubscriber*     getCurrSubscriber();
+    virtual void        getImageSize (int &w, int&h) = 0;
```
Comment (Floessie) [inline rtgui/editcallbacks.h:182]: I'm always wondering why we don't have a `Size` or `Dimension` class for things like this. Or do we? And if we had/have why such a size could be negative.

**Review by Floessie — COMMENTED** [review]
As you mentioned `editcallbacks.h` I had a look and made some uninformed comments. Some things need to be changed, others might not be changeable.

**Comment by Desmis** [comment]
@Pandagrapher 
Ce serait bien, merci :)

jacques

**Comment by Hombre57** [comment]
@Floessie Thanks for the review. As is, this branch is only a split of the actual one. I'll merge it as is so that merging into other branch will be easier, then I'll answer/address all your comments.

@Pandagrapher Can you pull the dev branch before Saturday evening ? I'll update this branch to enhance it based on @Floessie's recommendations.

File: rtgui/editcallbacks.h (hunk 1 above)
Comment (Hombre57) [reply rtgui/editcallbacks.h:42]: @Floessie `provider` is owned by the `EditDataProvider`, and setting this parameter through `setEditProvider` is connecting the two so they communicate, i.e. when you press an Edit button somewhere.

File: rtgui/editcallbacks.h (hunk 2 above)
Comment (Hombre57) [reply rtgui/editcallbacks.h:45]: @Floessie At first sight, yes, it should make things simpler to use a `std::unique_ptr<>`. I'll look at this.

File: rtgui/editcallbacks.h (hunk 3 above)
Comment (Hombre57) [reply rtgui/editcallbacks.h:47]: Done.

File: rtgui/editcallbacks.h (hunk 5 above)
Comment (Hombre57) [reply rtgui/editcallbacks.h:57]: Already answered above.

File: rtgui/editcallbacks.h (hunk 6 above)
Comment (Hombre57) [reply rtgui/editcallbacks.h:58]: `const` added.

File: rtgui/editcallbacks.h (hunk 7 above)
Comment (Hombre57) [reply rtgui/editcallbacks.h:60]: Done

File: rtgui/editcallbacks.h (hunk 8 above)
Comment (Hombre57) [reply rtgui/editcallbacks.h:64]: Done

File: rtgui/editcallbacks.h (hunk 9 above)
Comment (Hombre57) [reply rtgui/editcallbacks.h:72]: I don't know what `top level` means in this context, but I removed it.

File: rtgui/editcallbacks.h (hunk 6 above)
Comment (Hombre57) [reply rtgui/editcallbacks.h:58]: Reverting the modification... the `EditDataProvider` has to remain mutable.

File: rtgui/editcallbacks.h (hunk 2 above)
Comment (Hombre57) [reply rtgui/editcallbacks.h:45]: @Floessie I tried to convert to `std::unique_ptr<>`, however `std::vector` doesn't accept subclass of `Geometry`in when pushing back e.g. a `std::unique_ptr<Cicle>` . How do you handle that ?

**Comment by Desmis** [comment]
@Pandagrapher  @Hombre57 
I try to merge "dev" in "newlocallab" with the last modifications made to "edicallbacks"....
Many error messages in "controlspotpanel.h"and "controlspotpanel.cc"
==> mouseover   object  (private)
==> EditSubscriber::action = ES_ACTION_DRAGGING // or ES_ACTION_NONE;
Could you have a look

Thank you
jacques

File: rtgui/editcallbacks.h (hunk 2 above)
Comment (Floessie) [reply rtgui/editcallbacks.h:45]: @Hombre57 Eh? Works like a charm here...

```patch
diff --git a/rtgui/cropwindow.cc b/rtgui/cropwindow.cc
index 4b2c1a6cc..48bf5fad2 100644
--- a/rtgui/cropwindow.cc
+++ b/rtgui/cropwindow.cc
@@ -1838,23 +1838,23 @@ void CropWindow::expose (Cairo::RefPtr<Cairo::Context> cr)
                 }
 
                 // drawing Subscriber's visible geometry
-                const std::vector<Geometry*> visibleGeom = editSubscriber->getVisibleGeometry();
+                const auto& visibleGeom = editSubscriber->getVisibleGeometry();
                 cr->set_antialias(Cairo::ANTIALIAS_DEFAULT); // ANTIALIAS_SUBPIXEL ?
                 cr->set_line_cap(Cairo::LINE_CAP_SQUARE);
                 cr->set_line_join(Cairo::LINE_JOIN_ROUND);
 
                 // drawing outer lines
-                for (auto geom : visibleGeom) {
+                for (auto& geom : visibleGeom) {
                     geom->drawOuterGeometry(cr, this, *this);
                 }
 
                 // drawing inner lines
-                for (auto geom : visibleGeom) {
+                for (auto& geom : visibleGeom) {
                     geom->drawInnerGeometry(cr, this, *this);
                 }
 
                 // drawing to the "mouse over" channel
-                const auto mouseOverGeom = editSubscriber->getMouseOverGeometry();
+                const auto& mouseOverGeom = editSubscriber->getMouseOverGeometry();
                 if (mouseOverGeom.size()) {
                     if (mouseOverGeom.size() > 65534) {
                         // once it has been switched to OM_65535, it won't return back to OM_255
@@ -1876,7 +1876,7 @@ void CropWindow::expose (Cairo::RefPtr<Cairo::Context> cr)
                     crMO->fill();
 
                     int a=0;
-                    for (auto moGeom : mouseOverGeom) {
+                    for (auto& moGeom : mouseOverGeom) {
                         moGeom->drawToMOChannel(crMO, a, this, *this);
                         ++a;
                     }
diff --git a/rtgui/curveeditor.cc b/rtgui/curveeditor.cc
index fd41ba6e6..b5dd17182 100644
--- a/rtgui/curveeditor.cc
+++ b/rtgui/curveeditor.cc
@@ -449,7 +449,7 @@ void CurveEditor::switchOffEditMode ()
     EditSubscriber::switchOffEditMode();  // disconnect
 }
 
-const bool CurveEditor::mouseOver(const int modifierKey)
+bool CurveEditor::mouseOver(const int modifierKey)
 {
     EditDataProvider* provider = getEditProvider();
     subGroup->pipetteMouseOver(provider, modifierKey);
@@ -490,7 +490,7 @@ bool CurveEditor::drag1(const int modifierKey)
     return false;
 }
 
-const CursorShape CurveEditor::getCursor(const int objectID) const
+CursorShape CurveEditor::getCursor(const int objectID) const
 {
     if (remoteDrag) {
         return CSResizeHeight;
diff --git a/rtgui/curveeditor.h b/rtgui/curveeditor.h
index a4eeb0971..9d06ba22c 100644
--- a/rtgui/curveeditor.h
+++ b/rtgui/curveeditor.h
@@ -128,11 +128,11 @@ public:
     sigc::signal<void> signal_curvepoint_release();
 
     void switchOffEditMode () override;
-    const bool mouseOver(const int modifierKey) override;
+    bool mouseOver(int modifierKey) override;
     bool button1Pressed(const int modifierKey) override;
     bool button1Released() override;
     bool drag1(const int modifierKey) override;
-    const CursorShape getCursor(const int objectID) const override;
+    CursorShape getCursor(int objectID) const override;
 
 
 };
diff --git a/rtgui/editcallbacks.cc b/rtgui/editcallbacks.cc
index e130d7686..0a1b05abc 100644
--- a/rtgui/editcallbacks.cc
+++ b/rtgui/editcallbacks.cc
@@ -19,6 +19,8 @@
 
 #include "editcallbacks.h"
 
+#include "editwidgets.h"
+
 EditSubscriber::EditSubscriber (EditType editType) :
     ID(EUID_None),
     editingType(editType),
@@ -27,6 +29,8 @@ EditSubscriber::EditSubscriber (EditType editType) :
     action(EditSubscriber::Action::NONE)
 {}
 
+EditSubscriber::~EditSubscriber() = default;
+
 void EditSubscriber::setEditProvider(EditDataProvider *provider)
 {
     this->provider = provider;
@@ -38,7 +42,7 @@ void EditSubscriber::setEditID(EditUniqueID ID, BufferType buffType)
     bufferType = buffType;
 }
 
-const bool EditSubscriber::isCurrentSubscriber() const
+bool EditSubscriber::isCurrentSubscriber() const
 {
     //if (provider && provider->getCurrSubscriber())
     //  return provider->getCurrSubscriber()->getEditID() == ID;
@@ -69,27 +73,27 @@ void EditSubscriber::switchOffEditMode()
     unsubscribe();
 }
 
-const  EditUniqueID EditSubscriber::getEditID() const
+EditUniqueID EditSubscriber::getEditID() const
 {
     return ID;
 }
 
-const EditType EditSubscriber::getEditingType() const
+EditType EditSubscriber::getEditingType() const
 {
     return editingType;
 }
 
-const  BufferType EditSubscriber::getPipetteBufferType() const
+BufferType EditSubscriber::getPipetteBufferType() const
 {
     return bufferType;
 }
 
-const bool EditSubscriber::isDragging() const
+bool EditSubscriber::isDragging() const
 {
     return action == EditSubscriber::Action::DRAGGING;
 }
 
-const bool EditSubscriber::isPicking() const
+bool EditSubscriber::isPicking() const
 {
     return action == EditSubscriber::Action::PICKING;
 }
@@ -172,7 +176,7 @@ void EditDataProvider::setPipetteVal3(float newVal)
     pipetteVal3 = newVal;
 }
 
-const CursorShape EditDataProvider::getCursor(int objectID) const
+CursorShape EditDataProvider::getCursor(int objectID) const
 {
     if (currSubscriber) {
         currSubscriber->getCursor(objectID);
@@ -190,11 +194,11 @@ EditDataProvider* EditSubscriber::getEditProvider () {
     return provider;
 }
 
-const CursorShape EditSubscriber::getCursor (int objectID) const {
+CursorShape EditSubscriber::getCursor (int objectID) const {
     return CSHandOpen;
 }
 
-const bool EditSubscriber::mouseOver (const int modifierKey) {
+bool EditSubscriber::mouseOver (const int modifierKey) {
     return false;
 }
 
@@ -246,14 +250,16 @@ bool EditSubscriber::pick3 (const bool picked) {
     return false;
 }
 
-const std::vector<Geometry*>& EditSubscriber::getVisibleGeometry () {
+const std::vector<std::unique_ptr<Geometry>>& EditSubscriber::getVisibleGeometry()
+{
     return visibleGeometry;
 }
 
-const std::vector<Geometry*>& EditSubscriber::getMouseOverGeometry () {
+const std::vector<std::unique_ptr<Geometry>>& EditSubscriber::getMouseOverGeometry()
+{
     return mouseOverGeometry;
 }
 
-const int EditDataProvider::getPipetteRectSize () const {
+int EditDataProvider::getPipetteRectSize () const {
     return 8; // TODO: make a GUI
 }
diff --git a/rtgui/editcallbacks.h b/rtgui/editcallbacks.h
index 3ec15e00f..d11d261e4 100644
--- a/rtgui/editcallbacks.h
+++ b/rtgui/editcallbacks.h
@@ -61,19 +61,19 @@ public:
     void               setEditProvider(EditDataProvider *provider);
     EditDataProvider*  getEditProvider ();
     void               setEditID(EditUniqueID ID, BufferType buffType);
-    bool               isCurrentSubscriber();
+    bool               isCurrentSubscriber() const;
     virtual void       subscribe();
     virtual void       unsubscribe();
-    virtual void       switchOffEditMode ();    /// Occurs when the user want to stop the editing mode
-    EditUniqueID       getEditID();
-    EditType           getEditingType();
-    BufferType         getPipetteBufferType();
-    bool               isDragging();            /// Returns true if something is being dragged and drag events has to be sent (object mode only)
-    bool               isPicking();             /// Returns true if something is being picked
+    virtual void       switchOffEditMode();           /// Occurs when the user want to stop the editing mode
+    EditUniqueID       getEditID() const;
+    EditType           getEditingType() const;
+    BufferType         getPipetteBufferType() const;
+    bool               isDragging() const;            /// Returns true if something is being dragged and drag events has to be sent (object mode only)
+    bool               isPicking() const;             /// Returns true if something is being picked
 
     /** @brief Get the cursor to be displayed when above handles
     @param objectID object currently "hovered" */
-    virtual CursorShape getCursor (const int objectID);
+    virtual CursorShape getCursor (int objectID) const;
 
     /** @brief Triggered when the mouse is moving over an object
     This method is also triggered when the cursor is moving over the image in ET_PIPETTE mode
@@ -146,16 +146,10 @@ public:
     virtual bool pick3 (const bool picked);
 
     /** @brief Get the geometry to be shown to the user */
-    const std::vector<std::unique_ptr<Geometry>>& getVisibleGeometry()
-    {
-        return visibleGeometry;
-    }
+    const std::vector<std::unique_ptr<Geometry>>& getVisibleGeometry();
 
     /** @brief Get the geometry to be drawn in the "mouse over" channel, hidden from the user */
-    const std::vector<std::unique_ptr<Geometry>>& getMouseOverGeometry()
-    {
-        return mouseOverGeometry;
-    }
+    const std::vector<std::unique_ptr<Geometry>>& getMouseOverGeometry();
 };
 
 /** @brief Class to handle the furniture of data to the subscribers.
@@ -196,8 +190,8 @@ public:
     void setPipetteVal1(float newVal);
     void setPipetteVal2(float newVal);
     void setPipetteVal3(float newVal);
-    virtual const CursorShape getCursor(int objectID) const;
-    const int getPipetteRectSize () const;
+    virtual CursorShape getCursor(int objectID) const;
+    int getPipetteRectSize () const;
     EditSubscriber* getCurrSubscriber() const;
     virtual void getImageSize (int &w, int&h) = 0;
 };
diff --git a/rtgui/editwidgets.h b/rtgui/editwidgets.h
index ec935291e..3246d7377 100644
--- a/rtgui/editwidgets.h
+++ b/rtgui/editwidgets.h
@@ -538,5 +538,11 @@ inline Line::Line (int beginX, int beginY, int endX, int endY) :
         begin (beginX, beginY), end (endX, endY) {
 }
 
+#else
+
+class Geometry
+{
+};
+
 #endif
 
diff --git a/rtgui/gradient.cc b/rtgui/gradient.cc
index e4e006cbf..11a50b551 100644
--- a/rtgui/gradient.cc
+++ b/rtgui/gradient.cc
@@ -1,6 +1,8 @@
 /*
  *  This file is part of RawTherapee.
  */
+#include <array>
+
 #include "gradient.h"
 
 #include "editwidgets.h"
@@ -51,8 +53,6 @@ Gradient::Gradient () : FoldableToolPanel(this, "gradient", M("TP_GRADIENT_LABEL
     pack_start (*centerY, Gtk::PACK_SHRINK, 0);
 
     // Instantiating the Editing geometry; positions will be initialized later
-    Line *hLine, *vLine, *featherLine[2];
-    Circle *centerCircle;
 
     // Visible geometry
     std::unique_ptr<Line> hLine(new Line);
@@ -60,12 +60,11 @@ Gradient::Gradient () : FoldableToolPanel(this, "gradient", M("TP_GRADIENT_LABEL
     std::unique_ptr<Line> vLine(new Line);
     hLine->datum = vLine->datum = Geometry::IMAGE;
 
-    std::array<std::unique_ptr<Line>, 2> featherLine = {
-        new Line,
-        new Line
-    };
+    std::array<std::unique_ptr<Line>, 2> featherLine;
 
+    featherLine[0].reset(new Line);
     featherLine[0]->innerLineWidth = 2;
+    featherLine[1].reset(new Line);
     featherLine[1]->innerLineWidth = 2;
     featherLine[0]->datum = featherLine[1]->datum = Geometry::IMAGE;
 
@@ -87,12 +86,9 @@ Gradient::Gradient () : FoldableToolPanel(this, "gradient", M("TP_GRADIENT_LABEL
     vLine.reset(new Line);
     hLine->datum = vLine->datum = Geometry::IMAGE;
 
-    featherLine = {
-        new Line,
-        new Line
-    };
-
+    featherLine[0].reset(new Line);
     featherLine[0]->innerLineWidth = 2;
+    featherLine[1].reset(new Line);
     featherLine[1]->innerLineWidth = 2;
     featherLine[0]->datum = featherLine[1]->datum = Geometry::IMAGE;
 
@@ -113,13 +109,6 @@ Gradient::Gradient () : FoldableToolPanel(this, "gradient", M("TP_GRADIENT_LABEL
 
 Gradient::~Gradient()
 {
-    for (std::vector<Geometry*>::const_iterator i = visibleGeometry.begin(); i != visibleGeometry.end(); ++i) {
-        delete *i;
-    }
-
-    for (std::vector<Geometry*>::const_iterator i = mouseOverGeometry.begin(); i != mouseOverGeometry.end(); ++i) {
-        delete *i;
-    }
 }
 
 void Gradient::read (const ProcParams* pp, const ParamsEdited* pedited)
@@ -170,27 +159,27 @@ void Gradient::updateGeometry(const int centerX, const int centerY, const double
     const auto decay = feather * rtengine::norm2<double> (imW, imH) / 200.0;
     rtengine::Coord origin (imW / 2 + centerX * imW / 200, imH / 2 + centerY * imH / 200);
 
-    const auto updateLine = [&](Geometry* geometry, const float radius, const float begin, const float end)
+    const auto updateLine = [&](const std::unique_ptr<Geometry>& geometry, const float radius, const float begin, const float end)
     {
-        const auto line = static_cast<Line*>(geometry);
+        const auto line = static_cast<Line*>(geometry.get());
         line->begin = PolarCoord(radius, -degree + begin);
         line->begin += origin;
         line->end = PolarCoord(radius, -degree + end);
         line->end += origin;
     };
 
-    const auto updateLineWithDecay = [&](Geometry* geometry, const float radius, const float offSetAngle)
+    const auto updateLineWithDecay = [&](const std::unique_ptr<Geometry>& geometry, const float radius, const float offSetAngle)
     {
-        const auto line = static_cast<Line*>(geometry);
+        const auto line = static_cast<Line*>(geometry.get());
         line->begin = PolarCoord (radius, -degree + 180.) + PolarCoord (decay, -degree + offSetAngle);
         line->begin += origin;
         line->end = PolarCoord (radius, -degree) + PolarCoord (decay, -degree + offSetAngle);
         line->end += origin;
     };
 
-    const auto updateCircle = [&](Geometry* geometry)
+    const auto updateCircle = [&](const std::unique_ptr<Geometry>& geometry)
     {
-        const auto circle = static_cast<Circle*>(geometry);
+        const auto circle = static_cast<Circle*>(geometry.get());
         circle->center = origin;
     };
 
@@ -337,7 +326,7 @@ void Gradient::editToggled ()
     }
 }
 
-const CursorShape Gradient::getCursor(const int objectID) const
+CursorShape Gradient::getCursor(const int objectID) const
 {
     switch (objectID) {
     case (0):
@@ -363,7 +352,7 @@ const CursorShape Gradient::getCursor(const int objectID) const
     }
 }
 
-const bool Gradient::mouseOver(const int modifierKey)
+bool Gradient::mouseOver(const int modifierKey)
 {
     EditDataProvider* editProvider = getEditProvider();
 
diff --git a/rtgui/gradient.h b/rtgui/gradient.h
index 49799dc7b..8fbac6ece 100644
--- a/rtgui/gradient.h
+++ b/rtgui/gradient.h
@@ -52,8 +52,8 @@ public:
     void setEditProvider (EditDataProvider* provider) override;
 
     // EditSubscriber interface
-    const CursorShape getCursor(const int objectID) const override;
-    const bool mouseOver(const int modifierKey) override;
+    CursorShape getCursor(int objectID) const override;
+    bool mouseOver(int modifierKey) override;
     bool button1Pressed(const int modifierKey) override;
     bool button1Released() override;
     bool drag1(const int modifierKey) override;
```

**Comment by Floessie** [comment]
@Desmis `ES_ACTION_DRAGGING` is now `EditSubscriber::Action::DRAGGING` and the like.

**Comment by Floessie** [comment]
@Hombre57 Your latest changes introduce two problems:

1. `const` return values (like in `virtual const bool mouseOver(int modifierKey);`) which are ineffective for const correctness (the return value is copied anyway) but play a role in overriding.
2. DOS line endings.

Would be nice if that could be corrected for the 5.6 release. If you don't have the time just let me know and I'll care for those two points.

Best,
Flössie

**Comment by Desmis** [comment]
@Floessie 
Thank you for this information
Now I can compile but, in addition to this change, I have

replace  bool mouseOver(int modifierKey);  by   const bool mouseOver(int modifierKey);  in controlspotpanel.h and cc

change int "object" from "private" to" public" in editcallbacks.h, and other minor changes

jacques

**Comment by Desmis** [comment]
@Floessie 
I merge "newlocallab" with "dev" and commit the change
All seems to work :)
jacques

**Comment by Floessie** [comment]
@Desmis 
> change int "object" from "private" to" public" in editcallbacks.h, and other minor changes

I think I saw a `getObject()` and `setObject()` in @Hombre57's changes...

**Comment by Floessie** [comment]
@Hombre57 I picked up the DOS line endings from a different branch when I stashed the changes to transfer them to `dev`, so all is fine in this respect. Now I'll fix the wrong `const`s directly on `dev`...

**Comment by Hombre57** [comment]
@Floessie Yes, please cleanup this mess, I won't do that again, for sure, since you know where and when to use `const`, not me.

File: rtgui/editcallbacks.h (hunk 2 above)
Comment (Hombre57) [reply rtgui/editcallbacks.h:45]: @Floessie Well... I clearly loose my time in this project, this seems so easy for you. Could you please commit this patch ?

**Comment by Floessie** [comment]
@Hombre57 Sorry for the confusion and thanks for `editcallbacks.h`. :+1:

**Pull request merged**